tests/simulator_test.cpp

```diff
@@ -2,82 +2,269 @@
+  // collector radius. Thats why collision point its just the point between two
+  // collectors origin - height of the equilateral triangle.
+  Vec3 refCollision =
+      (OriginOfLastCollector + OriginOfPenultimateCollector) / 2 -
+      Vec3(0, 0, refCollectorRadius * std::sqrt(3) / 2);
+
+  ASSERT_EQ(refCollision, hitData.collisionPoint());
+
+  Ray straightDown(kVecZero, -kVecUp);
+  ASSERT_FALSE(performHitCollector(energyCollectors, straightDown, &hitData))
+      << "Collision Point: " << hitData.collisionPoint();
+
+  Ray alongX(kVecZero, kVecX);
+  ASSERT_TRUE(performHitCollector(energyCollectors, alongX, &hitData));
+
+  ASSERT_EQ(alongX.at(collectorPositionRadius - refCollectorRadius),
+            hitData.collisionPoint());
+
+  Ray alongY(kVecZero, kVecY);
+  ASSERT_TRUE(performHitCollector(energyCollectors, alongY, &hitData));
+
+  ASSERT_EQ(alongY.at(collectorPositionRadius - refCollectorRadius),
+            hitData.collisionPoint());
+
+  Ray at30(kVecZero, Vec3(std::cos(deg2rad(30)), 0, std::sin(deg2rad(30))));
+  ASSERT_TRUE(performHitCollector(energyCollectors, at30, &hitData));
+
+  Ray at30other(kVecZero,
+                Vec3(-std::cos(deg2rad(30)), 0, std::sin(deg2rad(30))));
+  ASSERT_TRUE(performHitCollector(energyCollectors, at30other, &hitData));
+
+  Ray at30XY(kVecZero, Vec3(std::cos(deg2rad(30)), std::cos(deg2rad(30)),
+                            std::sin(deg2rad(60))));
+  ASSERT_FALSE(performHitCollector(energyCollectors, at30XY, &hitData))
+      << "Collision Point: " << hitData.collisionPoint();
+
+  Ray at30XYOther(kVecZero, Vec3(-std::cos(deg2rad(30)), -std::cos(deg2rad(30)),
+                                 std::sin(deg2rad(60))));
+  ASSERT_FALSE(performHitCollector(energyCollectors, at30XYOther, &hitData))
+      << "Collision Point: " << hitData.collisionPoint();
+}
+
+TEST_F(EnergyCollectorTest, PreviousBuggedNearTopCollectorShouldHitOddNumber) {
+
+  const FakeModel nonEmptyModel(false);
+  const int numCollectors = 37;
+
+  auto energyCollectors = buildCollectors(nonEmptyModel, numCollectors);
+  ASSERT_EQ(energyCollectors.size(), numCollectors);
+
+  const float collectorPositionRadius = 4;
+  // this is how previous implementation was caclualating radius of energy
+  // collector
+  float invalidEnergyCollectorRadius =
+      2 * kPi * collectorPositionRadius / static_cast<float>(numCollectors);
+
+  RayHitData hitData;
+  Ray previousNotHit1(kVecZero, Vec3(0, 1.01 * invalidEnergyCollectorRadius,
+                                     collectorPositionRadius));
+  ASSERT_TRUE(performHitCollector(energyCollectors, previousNotHit1, &hitData));
+
+  Ray previousNotHit2(kVecZero, Vec3(0, -1.01 * invalidEnergyCollectorRadius,
+                                     collectorPositionRadius));
+
+  ASSERT_TRUE(performHitCollector(energyCollectors, previousNotHit2, &hitData));
+}
+
+TEST_F(EnergyCollectorTest, PreviousBuggedNearTopCollectorShouldHitEvenNumber) {
```

Comment:
Why do we care that it was not working? Just write what the test is about. E.g

"NoHoleAtPoleEvenCollectors"

Reply:
fixed both, for odd number of colectors and even number.